### PR TITLE
debug: add record type filter to debug keys

### LINKED ISSUE #64787: Add a debug command to introspect intents and transactions in storage
**Is your feature request related to a problem? Please describe.**
When debugging database files from production system in context of #60585 it is necessary to extract data about abandoned or active intents and correlate them to abandoned and active transactions. Extracting this from existing commands is very compute intensive and hard to make for many nodes.

**Describe the solution you'd like**
By adding specific filters to debug keys command or providing a separate debug command we could make investigation process more effective.

**Describe alternatives you've considered**
Parsing through output of keys or range-data is too slow.

### PR DESCRIPTION
**Pull request by aliher1911**
Previously debug keys could only print storage keys that belong to a range.
To investigate transaction commit anomalies it is helpful to extract data about
intents, transactions and keys separately as number of keys could be too big
to process.
To address this, new command line flag --type is added that allows selecting
only single key type to dump (values, intents or txns).

Release note (cli change): --type option to debug keys

### CODE REVIEW

File: pkg/cli/debug.go
Comment (erikgrinaker) [inline pkg/cli/debug.go:251]: I think we should check that there's no timestamp as well (intents do not have timestamps as they're not written yet).

File: pkg/cli/cliflags/flags.go
Comment (erikgrinaker) [inline pkg/cli/cliflags/flags.go:1037]: Wrap at 80 chars.

File: pkg/cli/flags.go
Comment (erikgrinaker) [inline pkg/cli/flags.go:238]: Maybe drop the `show` prefix here.

File: pkg/cli/debug.go
Comment (erikgrinaker) [inline pkg/cli/debug.go:20]: This should be grouped with other non-stdlib imports below.

File: pkg/cli/flags.go
Comment (erikgrinaker) [inline pkg/cli/flags.go:232]: I wonder if we maybe should just use a `string` base type instead of `int8` for these. That would make `String()` and `Set()` trivial.

**Review by erikgrinaker — COMMENTED** [review]
Looks good! I tried running `cockroach debug keys` (from `master`) on some customer data, but it just panics:

```
panic: iterator with constraint=2 is being used with key /Local/RangeID/51"" that has constraint=1
```

Maybe we'll need to fix this as well? Or am I doing something wrong?

**Review by aliher1911 — COMMENTED** [review]
When we list transactions, I think we could just iterate over local range which makes it run pretty quick, but if we use full range then it could be like 15 min to scan the full set of files even if we know that there should be no transaction records there. We could limit it explicitly by providing --from/--to arguments but one has to keep that in mind. Alternatively we could apply default range based on type of records we are listing if no range was provided explicitly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-:-M_Pvx0R2enkODuv-W6h:bdnq8vp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @erikgrinaker)

___
*[pkg/cli/debug.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E6-r1-20:-M_Puv1F5S8drLes5dQm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/debug.go#L20)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This should be grouped with other non-stdlib imports below.
</blockquote></details>

Done.
___
*[pkg/cli/debug.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E6-r1-251:-M_Puw164UPw7Eqmd9R9:b-95rjv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/debug.go#L251)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I think we should check that there's no timestamp as well (intents do not have timestamps as they're not written yet).
</blockquote></details>

isValue() above is checking exactly that so we don't need to check again. The Txn check here is to catch cases where we successfully deserialize some other payload into empty meta which is causing lots of false positives.
___
*[pkg/cli/cliflags/flags.go, line 1034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E4-r1-1034:-M_PvPwzEPFFI37O8_5N:b2qm3o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/cliflags/flags.go#L1034)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Maybe `filter` or `type` or something like that?
</blockquote></details>

Type seems reasonable. I decided against filter initially as it was opposite to what we do imho.

**Review by erikgrinaker — COMMENTED** [review]
>I looked on it and constraint is which keys iterator is iterating - 2 means global, 1 local. So iterator is constrained to global keys but iterating starting from local and that causes panic. It is happening inside intent_interleaving_iter which is relatively new i believe so might be a bug in how debug keys uses it. Did you provide any key ranges for it btw?

Right. I don't provide any ranges, no. Does local mean range-local? If so, we'll need to scan those as well, since txn entries are range-local I think. Maybe we can just drop the constraint completely?

>When we list transactions, I think we could just iterate over local range which makes it run pretty quick, but if we use full range then it could be like 15 min to scan the full set of files even if we know that there should be no transaction records there. We could limit it explicitly by providing --from/--to arguments but one has to keep that in mind. Alternatively we could apply default range based on type of records we are listing if no range was provided explicitly.

Applying a default range for txns makes sense, since we know where they're stored.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-:-M_PxVbHBGk_qNITTRNq:b-c2uhay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

___
*[pkg/cli/debug.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E6-r1-251:-M_Py0sCC1hcwqlT2iRK:bo43g42) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/debug.go#L251)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

isValue() above is checking exactly that so we don't need to check again. The Txn check here is to catch cases where we successfully deserialize some other payload into empty meta which is causing lots of false positives.
</blockquote></details>

Ah, you're right, my bad.

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/cli/flags.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E7-r1-232:-M_Q4HPr2t5izQ5cOOfl:bo701wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/flags.go#L232)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I wonder if we maybe should just use a `string` base type instead of `int8` for these. That would make `String()` and `Set()` trivial.
</blockquote></details>

I was actually taking inspiration from other flags like engine type and they don't use string. Using string only marginally changes the plumbing as it still needs to be casted back and forth and checked against list of allowed values.
___
*[pkg/cli/flags.go, line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E7-r1-238:-M_QEpmrCm9B7boV_0Q2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/flags.go#L238)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Maybe drop the `show` prefix here.
</blockquote></details>

Done.
___
*[pkg/cli/cliflags/flags.go, line 1037 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E4-r1-1037:-M_QFhuv4dNcCL9TU1nn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/cliflags/flags.go#L1037)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Wrap at 80 chars.
</blockquote></details>

Done.

**Review by erikgrinaker — COMMENTED** [review]
> To address this, new command line flag show-keys is added that allows selecting
only single key type to dump.

This needs to be updated with the actual command.

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-:-M_QRI-z4WR1gy_rtakz:b-eq2jae)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

___
*[pkg/cli/flags.go, line 232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_6LCaDg2S3z1FVR6E7-r1-232:-M_QRmoP2N-2k7PaTnNV:bsvgpv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/797a6c392b4abb621a1ed5977b71b0ab5b713cbf/pkg/cli/flags.go#L232)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

I was actually taking inspiration from other flags like engine type and they don't use string. Using string only marginally changes the plumbing as it still needs to be casted back and forth and checked against list of allowed values.
</blockquote></details>

Got it, let's follow the existing convention.
___
*[pkg/cli/cliflags/flags.go, line 1033 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_QS1CpBDBOQb9OJjC3:-M_QS1CpBDBOQb9OJjC4:bsf5v90) ([raw file](https://github.com/cockroachdb/cockroach/blob/f523e0a61fe0bf7b1e05c73a2651f896f9a38f44/pkg/cli/cliflags/flags.go#L1033)):*
> ```Go
> 	}
> 
> 	FilterKeys = FlagInfo{
> ```

Now that this flag is `type`, we should use "type" as the term throughout, i.e. rename `FilterKeys` to `Type` or `KeyType` or something, and the same for the `filter` type and `show` prefixes.

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/cli/cliflags/flags.go, line 1033 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_QS1CpBDBOQb9OJjC3:-M_QnIIg8dLBTgnS9VPe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f523e0a61fe0bf7b1e05c73a2651f896f9a38f44/pkg/cli/cliflags/flags.go#L1033)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Now that this flag is `type`, we should use "type" as the term throughout, i.e. rename `FilterKeys` to `Type` or `KeyType` or something, and the same for the `filter` type and `show` prefixes.
</blockquote></details>

Done.

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/cli/cliflags/flags.go, line 1033 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-M_QS1CpBDBOQb9OJjC3:-M_QnQLV1FvjMTOu-IIn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f523e0a61fe0bf7b1e05c73a2651f896f9a38f44/pkg/cli/cliflags/flags.go#L1033)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

**Review by erikgrinaker — APPROVED** [review]
:lgtm: But maybe get a second review from the CLI team as well.

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64879#-:-M_QqzM_9dWxxwyjzsYQ:b-u9w7s5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911)

**Comment by erikgrinaker** [comment]
Tried this out on some customer data I've been looking at, works great and completed much more quickly! 🚀

PS. You may want to CC the KV team on this.

**Review by knz — APPROVED** [review]
cli changes LGTM

**Comment by knz** [comment]
The release note can be improved though - it should be a full sentence and explain what is going on. For example:

```
Release note (cli change): The `cockroach debug XXXX` command now recognizes a new flag `--type`
that filters the entries displayed. This enables the user to focus on just one type which
simplifies certain troubleshooting scenarios.
```

**Comment by knz** [comment]
fyi bilal is off on PTO this weel

**Comment by aliher1911** [comment]
bors r=erikgrinaker,knz

**Pull request merged**
